pkgs/cli_util/lib/src/base_directories.dart

```diff
@@ -0,0 +1,274 @@
+  /// important enougn, not small enough, or not portable enough to store in
+  /// [dataHome]. For example: logs and indices.
+  ///
+  /// The directory location depends on the current [Platform.operatingSystem]:
+  /// - on **Windows**:
+  ///   - `%LOCALAPPDATA%\<tool>`
+  /// - on **Mac OS**:
+  ///   - `$HOME/Library/Application Support/<tool>`
+  /// - on **Linux**:
+  ///   - `$XDG_STATE_HOME/<tool>` if `$XDG_STATE_HOME` is defined, and
+  ///   - `$HOME/.local/state/<tool>` otherwise.
+  ///
+  /// The directory won't be created, the method merely returns the recommended
+  /// location.
+  ///
+  /// On some platforms, this path may overlap with [cacheHome], and
+  /// [runtimeHome].
+  ///
+  /// Throws an [EnvironmentNotFoundException] if the necessary environment
+  /// variables are undefined.
+  String stateHome() =>
+      path.join(_baseDirectory(_XdgBaseDirectoryType.state), tool);
+
+  String _baseDirectory(_XdgBaseDirectoryType dir) {
+    if (Platform.isWindows) {
+      return _baseDirectoryWindows(dir);
+    }
+    if (Platform.isMacOS) {
+      return _baseDirectoryMacOs(dir);
+    }
+    if (Platform.isLinux) {
+      return _baseDirectoryLinux(dir);
+    }
+    throw UnsupportedError('Unsupported platform: ${Platform.operatingSystem}');
+  }
+
+  String _baseDirectoryWindows(_XdgBaseDirectoryType dir) => switch (dir) {
+        _XdgBaseDirectoryType.config ||
+        _XdgBaseDirectoryType.data =>
+          _requireEnv('APPDATA'),
+        _XdgBaseDirectoryType.cache ||
+        _XdgBaseDirectoryType.runtime ||
+        _XdgBaseDirectoryType.state =>
+          _requireEnv('LOCALAPPDATA'),
+      };
+
+  String _baseDirectoryMacOs(_XdgBaseDirectoryType dir) => switch (dir) {
+        _XdgBaseDirectoryType.config =>
+          path.join(_home, 'Library', 'Preferences'),
+        _XdgBaseDirectoryType.data ||
+        _XdgBaseDirectoryType.state =>
+          path.join(_home, 'Library', 'Application Support'),
+        _XdgBaseDirectoryType.cache => path.join(_home, 'Library', 'Caches'),
+        _XdgBaseDirectoryType.runtime =>
+          // https://stackoverflow.com/a/76799489
+          path.join(_home, 'Library', 'Caches', 'TemporaryItems'),
+      };
+
+  String _baseDirectoryLinux(_XdgBaseDirectoryType dir) {
+    if (Platform.isLinux) {
+      final xdgEnv = switch (dir) {
+        _XdgBaseDirectoryType.config => 'XDG_CONFIG_HOME',
+        _XdgBaseDirectoryType.data => 'XDG_DATA_HOME',
+        _XdgBaseDirectoryType.state => 'XDG_STATE_HOME',
+        _XdgBaseDirectoryType.cache => 'XDG_CACHE_HOME',
+        // Note, DIR instead of HOME.
+        _XdgBaseDirectoryType.runtime => 'XDG_RUNTIME_DIR',
+      };
+      final envVar = environment[xdgEnv];
+      if (envVar != null) {
+        return envVar;
+      }
+    }
+
+    switch (dir) {
+      case _XdgBaseDirectoryType.runtime:
+      // Applications should print a fallback message ideally.
```

Comment:
About what, and how would they know.

This case falls through to the `cache` case. If we want the application to know, we should return `null` here, and let `runtimeHome` be nullable, and then tell users that if `runtimeHome` is `null`, they should use `cacheHome` instead *and notify the user*.

Then the user will do `baseDirs.runtimeHome ?? baseDirs.configHome()` and never report anything, but at least they have the option.

Reply:
It's unfortunate that complexity from the XDG specification leaks into the API
this way. I had hoped we could abstract some of it away.